src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -73,6 +73,11 @@ public abstract class NodeModel : ModelBase, IRenderPackageSource<NodeModel>, ID
         /// </summary>
         public virtual string CreationName { get { return this.Name; } }
 
+        /// <summary>
+        /// Returns all the upstream nodes for the given node
+        /// </summary>
+        public HashSet<NodeModel> Upstream = new HashSet<NodeModel>();
```

Comment:
perhaps we need to add that method `ComputeAllUpstreamNodes()` on the workspace that finds the roots of the graph and calls `ComputeUpstream` without recomputing the same node twice - I think it's fine if we just clarify this for now with a longer comment, and I can attempt this method in the next PR we discussed.